### PR TITLE
factorio: 0.12.29 headless + server module

### PR DESCRIPTION
**Pull request by elitak**
I bought this game through steam, so I can't test the authenticated download, nor get the sha256 sums for the tarballs. I need @Baughn or someone else to add and test them. I moved around the non-headless setup code, but tried to leave it intact, so please review that as well.

### CODE REVIEW

**Comment by Baughn** [comment]
This mixes the version bump, module creation and refactoring. Those should really each be in different commits.

I'll take a deeper look tonight sometime.

**Comment by elitak** [comment]
I got the hashes. Installation works. I'll test next and split into 3 commits.

**Comment by elitak** [comment]
I've briefly tested the x86 alpha client and it runs. Can't be 100% sure sound works because I only have a VM to test with, but an error dialog went away when i installed pulseaudio.

Did you mean that I should break this up into 3 pull requests or just commits?

**Comment by elitak** [comment]
Both alpha builds work with sound tested. Should be ready for final review + commit now.

**Comment by Baughn** [comment]
Three commits. Having them all in this PR is fine. Let me see, now...

Not *quite*. :P

File: nixos/modules/services/games/factorio.nix [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          The name of the savegame that will be used by the server.
+          When not present in ${stateDir}/saves, it will be created before starting the service.
+        '';
+      };
+      configFile = mkOption {
+        type = types.path;
+        default = configFile;
+        defaultText = "configFile";
+        description = ''
+          FIXME
```
Comment (Baughn) [inline nixos/modules/services/games/factorio.nix]: Yes, please do describe.

File: pkgs/games/factorio/default.nix [hunk 2]
```diff
@@ -4,80 +4,108 @@
...
+  authenticatedFetch = callPackage ./fetch.nix { inherit username password; };
 
-  variants = {
+  fetch = rec {
+    url = "https://www.factorio.com/get-download/${version}/${releaseType}/${arch.inUrl}";
+    name = "factorio_${releaseType}_${arch.inTar}-${version}.tar.gz"; # TODO take this from 302 redirection somehow? fetchurl doesn't help.
     x64 = {
-      url = "https://www.factorio.com/get-download/${version}/alpha/linux64";
-      sha256 = "01si5n9hb2h0c5q8k3hr3nphsakp9kki84qyp70dgddwqsn8wfjl";
+      alpha = authenticatedFetch { inherit      url; sha256 = "0vngfrjjib99k6czhg32rikfi36i3p3adx4mxc1z8bi5n70dbwqb"; };
```
Comment (Baughn) [inline pkgs/games/factorio/default.nix]: Try to keep lines below 80 characters when feasible. You can split these across multiple lines.

File: pkgs/games/factorio/default.nix [hunk 3]
```diff
@@ -4,80 +4,108 @@
...
+    cp -a data $out/share/factorio
+    cp -a bin/${arch.inTar}/factorio $out/bin/factorio
     patchelf \
       --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
-      $out/bin/factorio.${arch}
+      $out/bin/factorio
 
-    makeWrapper $out/bin/factorio.${arch} $out/bin/factorio \
+  '' + optionalString (! isHeadless) (''
+    mv $out/bin/factorio{,.${arch.inTar}}
```
Comment (Baughn) [inline pkgs/games/factorio/default.nix]: I will admit, this is clever. Maybe a bit too much.

Comment (elitak) [inline pkgs/games/factorio/default.nix]: The braces? or moving the non-headless path? or just the optionalString part?

The other approach for separating the derivations that I considered was putting "// if (! isHeadless { ... } else {}" at the end.

**Comment by joachifm** [comment]
Is this ready to go?

File: pkgs/games/factorio/default.nix (hunk 3 above)
Comment (Baughn) [inline pkgs/games/factorio/default.nix]: I meant the $out/bin/factorio{,.${arch.inTar}} part. You're generating two arguments to mv, but it looks like a single one to a casual glance.

I'd prefer to have two visually separate arguments.

Comment (elitak) [inline pkgs/games/factorio/default.nix]: ok, done

**Comment by Baughn** [comment]
From my perspective, yes it is, assuming it works. It seems you'll need to rebase again.

**Pull request merged**